[Compatibility] Implement new MatchData methods for 3.2

rwstauner opened the pull request:
This implements `#byteoffset`, `#deconstruct`, and `#deconstruct_keys` in `MatchData` to get the specs passing.

I added the `!inBounds` specialization to the existing byte begin/end methods for consistency with the (char) begin/end methods.  This got the error messages to match in the spec:
https://github.com/oracle/truffleruby/blob/ef93bdc18593f10dc4c1733e1af7a0d7a3bbe1b4/spec/ruby/core/matchdata/byteoffset_spec.rb#L80-L84

I took the two existing code blocks for "convert string/symbol/int to backref" and moved that to a new method.
I had to add an additional check to it to make the specs pass for an expected error:
https://github.com/oracle/truffleruby/blob/ef93bdc18593f10dc4c1733e1af7a0d7a3bbe1b4/spec/ruby/core/matchdata/byteoffset_spec.rb#L68

In the end I only needed that logic one additional time, so if it's better to have it inlined I can put it back how it was.
However the check I added does seem relevant for the begin/end methods (according to the new specs I added).

I changed `Truffle::Type.coerce_to(index, Integer, :to_int)` to `Primitive.rb_to_int(index)` to get the error message to match what the specs were expecting.  (I saw a reference to that in [some docs](https://github.com/oracle/truffleruby/blob/ef93bdc18593f10dc4c1733e1af7a0d7a3bbe1b4/doc/contributor/how-to-guide.md#L768-L771), but not sure what differences there might be.)

I was able to remove 3 excludes from the MRI Regexp tests.

rwstauner commented on src/main/ruby/truffleruby/core/match_data.rb:
```diff
@@ -61,6 +66,25 @@ def string
+
+    array_of_names.each do |key|
+      Truffle::Type.rb_check_type(key, Symbol)
+      break unless hash.key?(key)
+      ret[key] = hash[key]
```
I considered
```suggestion
      ret[key] = hash.fetch(key) { return ret }
```
I dislike doing the key lookup twice, but I'm not sure if that is better than introducing a block.

eregon replied:
There is `Primitive.hash_get_or_undefined` which could be used for that, or `hash.fetch(key, undefined)`.

eregon replied:
It makes sense to use `Primitive.hash_get_or_undefined` here, I or Andrii will do it to avoid an extra round-trip.

andrykonchin commented:
> I changed Truffle::Type.coerce_to(index, Integer, :to_int) to Primitive.rb_to_int(index) to get the error message to match what the specs were expecting. (I saw a reference to that in [some docs](https://github.com/oracle/truffleruby/blob/ef93bdc18593f10dc4c1733e1af7a0d7a3bbe1b4/doc/contributor/how-to-guide.md#L768-L771), but not sure what differences there might be.)

AFAIK `Primitive.rb_to_int` is kind of a specialised version of `Truffle::Type.rb_convert_type`.

Wondering if there is any performance benefits from using `Primitive.rb_to_int` cc @eregon.

andrykonchin commented on CHANGELOG.md:
```diff
@@ -29,6 +29,7 @@ Compatibility:
 * Set `RbConfig::CONFIG['configure_args']` for openssl and libyaml (#3170, #3303, @eregon).
 * Support `Socket.sockaddr_in(port, Socket::INADDR_ANY)` (#3361, @mtortonesi).
 * Implement the `Data` class from Ruby 3.2 (#3039, @moste00, @eregon).
+* Implement `MatchData` methods `byteoffset`, `deconstruct`, and `deconstruct_keys` from Ruby 3.2 (@rwstauner).
```
```suggestion
* Implement `MatchData` methods `byteoffset`, `deconstruct`, and `deconstruct_keys` from Ruby 3.2 (#3039, @rwstauner).
```

nirvdrum replied:
I'd also change to `MatchData#{byteoffset,deconstruct,deconstruct_key}` to be consistent with the rest of the changelog.

nirvdrum commented on spec/ruby/core/matchdata/begin_spec.rb:
```diff
@@ -36,6 +36,18 @@
       match_data.begin(obj).should == 2
     end
+
+    it "raises IndexError if index is out of matches" do
```
I was going to say "is out of matches" is awkward phrasing, but I see it matches the error message. Upstream recently fixed some message phrasing recently; maybe they'll be open to this one as well.

eregon replied:
I'd say let's use `out of bounds` in this spec description because that seems clearer, we generally don't use exception message in spec descriptions.

nirvdrum commented on spec/ruby/core/matchdata/begin_spec.rb:
```diff
@@ -68,6 +80,14 @@
       match_data.begin("æ").should == 1
     end
+
+    it "raises IndexError if there is no group with provided name" do
```
```suggestion
    it "raises IndexError if there is no group with the provided name" do
```

nirvdrum commented on spec/ruby/core/matchdata/begin_spec.rb:
```diff
@@ -100,5 +120,13 @@
       match_data.begin(:æ).should == 1
     end
+
+    it "raises IndexError if there is no group with provided name" do
```
```suggestion
    it "raises IndexError if there is no group with the provided name" do
```

nirvdrum commented on src/main/java/org/truffleruby/core/regexp/MatchDataNodes.java:
```diff
@@ -627,6 +627,14 @@ Object byteBegin(RubyMatchData matchData, int index,
             }
         }
 
+        @TruffleBoundary
```
I wonder if it's worth skipping `StringUtils.format` here and doing away with the boundary. This string can be trivially built with the `+` operator.

To be clear, I'm not saying you've done anything wrong here. I see you've matched what we're doing elsewhere in the file. And it's an exceptional case, so having the boundary isn't likely to cause problems. It's just that we can also remove it with minimal effort.

@eregon Thoughts on that?

eregon replied:
I think it's not worth spending time to optimize this likely-rare exceptional case.
And also not having the boundary would mean more code to PE.
Java String + is also not great to PE (I think can lead to some invoke dynamic stuff),
but the real killer here would be Integer#toString() which is definitely bad to PE (increase warmup, increase image size, unlikely to be faster PE'd than run behind a boundary).

However, this `@TruffleBoundary` on the specialization should be removed, because it hides the `throw` and prevents to PE the RaiseException constructor, getContext(), etc.
There is already a `@TruffleBoundary` on `StringUtils.format(` so that's enough.

nirvdrum commented on src/main/ruby/truffleruby/core/match_data.rb:
```diff
@@ -153,6 +163,20 @@ def match_length(n)
+  def backref_from_arg(index)
+    if Primitive.is_a?(index, String) || Primitive.is_a?(index, Symbol)
+      names_to_backref = Hash[Primitive.regexp_names(self.regexp)]
+      unless names_to_backref.key?(index.to_sym)
```
I'm finding this rather difficult to read. There's a lot going on. Can physically separate the `unless` statement block from the lines above and below with vertical space?

rwstauner replied:
I rewrote it a bit, does that improve readability?

eregon commented:
> Wondering if there is any performance benefits from using `Primitive.rb_to_int` cc @eregon.

It avoids a method call by being a Primitive (and so a node directly in the "caller" AST), plus if one call site needs actual coercion with `to_int` it won't affect any other call site, and this without needing splitting (which would be expensive for this).

eregon commented on src/main/java/org/truffleruby/core/regexp/MatchDataNodes.java:
```diff
@@ -649,6 +657,14 @@ Object byteEnd(RubyMatchData matchData, int index,
             }
         }
 
+        @TruffleBoundary
```
No @TruffleBoundary here

eregon commented on src/main/ruby/truffleruby/core/match_data.rb:
```diff
@@ -153,6 +163,21 @@ def match_length(n)
+
+  def backref_from_arg(index)
+    if Primitive.is_a?(index, String) || Primitive.is_a?(index, Symbol)
+      names_to_backref = Hash[Primitive.regexp_names(self.regexp)]
```
Creating a Hash to just lookup once/twice in it seems expensive, but it was already like that.
Probably we should add a primitive taking a name and returning the backrefs or last backref.
But the code was already like that, so let's keep as-is in this PR.

eregon reviewed (commented):
Looks good, please address my comments and then let's merge it.

eregon commented on src/main/ruby/truffleruby/core/match_data.rb:
```diff
@@ -153,6 +163,21 @@ def match_length(n)
+      names_to_backref = Hash[Primitive.regexp_names(self.regexp)]
+      key = index.to_sym
+
+      raise IndexError, "undefined group name reference: #{index}" unless names_to_backref.key?(key)
+
+      return names_to_backref[key].last
```
This also does two lookups.
Here it seems trivial to avoid with `array = h[k]; raise unless array` since `nil` is not a valid value in that `Hash[Symbol => Array]`

rwstauner replied:
Ah, quite right!  I started by wondering if ruby had a function like go's map fetch to tell you if the key existed and then got hung up checking for existence when it's obvious we were never expecting a nil there so it wasn't necessary.

rwstauner commented:
Removed the 2 TruffleBoundary annotations for begin/end byte, reworded those spec titles, dropped the `.key?` call, then rebased to avoid the CHANGELOG conflict

eregon commented:
> then rebased to avoid the CHANGELOG conflict

You don't need to worry about that one, it's a GitHub UI bug to say there is a conflict but there is never one because of the union merge strategy.
